### PR TITLE
[TMVA][SOFIE] ONNX parser with registry like pattern

### PR DESCRIPTION
**Pull request by axmat**
## Changes or fixes:

Use registry pattern for the ONNX parser. Now the operators and the tensor types are stored in a std::unordered_map. All the implemented operators are registered automatically when creating an object of type RModelParser_ONNX. When parsing an onnx graph the parser checks the type of the onnx node and throw an exception if it's not implemented or call the corresponding function to parse it.

This PR also makes SOFIE more modular. Now operators can be added simply by defining an operator class that inherits from ROperator, defining a function to parse the onnx node and registering this function to the parser.

## Example of adding a new ONNX operator

```{c++}
#include "onnx_proto3.pb.h"
#include "TMVA/RModel.hxx"
#include "TMVA/RModelParser_ONNX.hxx"

using namespace TMVA::Experimental::SOFIE;

template<typename T>
struct ROperator_Sign : public ROperator {
   std::string fNX;
   std::string fNY;
   std::vector<size_t> fShapeX;
   std::vector<size_t> fShapeY;

   ROperator_Sign(std::string nameX, std::string nameY) : fNX(nameX), fNY(nameY) {}

   std::vector<std::vector<size_t>> ShapeInference(std::vector<std::vector<size_t>> input) override {
      return input;
   }

   std::vector<ETensorType> TypeInference(std::vector<ETensorType> input) override {
      return input;
   }

   void Initialize(RModel& model) override {
        if (!model.CheckIfTensorAlreadyExist(fNX)) {
         throw std::runtime_error("TMVA::SOFIE - Tensor " + fNX + " not found.");
      }
      fShapeX = model.GetTensorShape(fNX);
      fShapeY = fShapeX;
      //fShapeY = ShapeInference({fShapeX})[0];
      model.AddIntermediateTensor(fNY, model.GetTensorType(fNX), fShapeY);
   }

   std::string Generate(std::string OpName) override {
      OpName = "op_" + OpName;
      std::stringstream out;

      out << SP << "\n//---- Operator" << OpName << "\n";
      size_t length = ConvertShapeToLength(fShapeX);
      out << SP << "for (size_t i = 0; i < " << length << "; i++) {\n";
      out << SP << SP << "if (tensor_" << fNX << "[i] > 0.) {\n";
      out << SP << SP << SP << "tensor_" << fNY << "[i] = 1.;\n";
      out << SP << SP << "} else if (tensor_" << fNX << "[i] < 0.) {\n";
      out << SP << SP << SP << "tensor_" << fNY << "[i] = -1.;\n";
      out << SP << SP << "} else {\n";
      out << SP << SP << SP << "tensor_" << fNY << "[i] = 0.;\n";
      out << SP << SP << "}\n";
      out << SP << "}\n";
      return out.str();
   }
};

// Parser for onnx::NodeProto of type Sign
ParserFuncSignature ParseSign = [](RModelParser_ONNX& parser, const onnx::NodeProto& nodeproto)
-> std::unique_ptr<ROperator> {
   std::unique_ptr<ROperator> op;

   ETensorType input_type = ETensorType::UNDEFINED;
   const std::string input_name = nodeproto.input(0);
   if (parser.IsRegisteredTensorType(input_name)) {
      input_type = parser.GetTensorType(input_name);
   } else {
      throw std::runtime_error("TMVA::SOFIE ONNX Parser Sign op has input tensor " + input_name +
                               " but its type is not yet registered");
   }

   const std::string output_name = nodeproto.output(0);
   switch (input_type) {
   case ETensorType::FLOAT: op.reset(new ROperator_Sign<float>(input_name, output_name)); break;
   default:
      throw std::runtime_error("TMVA::SOFIE - Unsupported - Sign Operator does not support imput type " +
                               std::to_string(static_cast<int>(input_type)));
   }

   if (!parser.IsRegisteredTensorType(output_name)) {
      parser.RegisterTensorType(output_name, input_type);
   }
   return op;
};

void SofieSign() {

   RModelParser_ONNX parser;

   // Register the Sign operator
   parser.RegisterOperator("Sign", ParseSign);

   // Parse the model
   RModel model = parser.Parse("Sign.onnx");
   model.Generate();
   model.OutputGenerated();
}
```
the ``onnx_proto3.pb.h`` is generated from the ``onnx_proto3`` file by runing ``protoc --cpp_out=. onnx_proto3``.

The generated code looks like this:
```{c++}
//Code generated automatically by TMVA for Inference of Model file [Sign.onnx] at [Sat Oct  8 13:10:40 2022] 

#ifndef TMVA_SOFIE_SIGN
#define TMVA_SOFIE_SIGN

#include<vector>
#include "TMVA/SOFIE_common.hxx"
#include <fstream>

namespace TMVA_SOFIE_Sign{
struct Session {
std::vector<float> fTensor_Y = std::vector<float>(5);
float * tensor_Y = fTensor_Y.data();


Session(std::string filename ="") {
   if (filename.empty()) filename = "Sign.dat";
}

std::vector<float> infer(float* tensor_X){
   
//---- Operatorop_0
   for (size_t i = 0; i < 5; i++) {
      if (tensor_X[i] > 0.) {
         tensor_Y[i] = 1.;
      } else if (tensor_X[i] < 0.) {
         tensor_Y[i] = -1.;
      } else {
         tensor_Y[i] = 0.;
      }
   }
   std::vector<float> ret (tensor_Y, tensor_Y + 5);
   return ret;
}
};
} //TMVA_SOFIE_Sign

#endif  // TMVA_SOFIE_SIGN
```

## Checklist:

- [x] tested changes locally

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2204/cxx17, ROOT-ubuntu2004/default, ROOT-ubuntu18.04/nortcxxmod

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2204/cxx17, ROOT-ubuntu2004/default, ROOT-ubuntu18.04/nortcxxmod

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2204/default, ROOT-ubuntu2004/cxx17, ROOT-ubuntu18.04/nortcxxmod

File: tmva/sofie_parsers/src/RModelParser_ONNX.cxx [hunk 1]
```diff
@@ -6,1658 +6,212 @@
...
-   std::unique_ptr<ROperator> op;
-   float fepsilon = 1e-05;
-	float fmomentum = 0.9;
-	std::size_t ftraining_mode = 0;
-
-   switch(input_type) {
-      case ETensorType::FLOAT:
-         if (nodeproto.input_size() == 5) {
-            op.reset(new ROperator_BatchNormalization<float>(fepsilon, fmomentum, ftraining_mode, nodeproto.input(0), nodeproto.input(1), nodeproto.input(2), nodeproto.input(3), nodeproto.input(4), nodeproto.output(0)));
+   // auto find = mapOptypeOperator.find(nodeproto.op_type());
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/RModelParser_ONNX.cxx:122]: Can we remove the commented code if they aren't required anymore?

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2204/default, ROOT-ubuntu2004/cxx17, ROOT-ubuntu18.04/nortcxxmod

File: tmva/sofie_parsers/src/RModelParser_ONNX.cxx (hunk 1 above)
Comment (axmat) [reply tmva/sofie_parsers/src/RModelParser_ONNX.cxx]: Done. Thank you @sanjibansg for the review!

**Comment by lmoneta** [comment]
@phsft-bot build just on ROOT-ubuntu2004/default, ROOT-ubuntu18.04/nortcxxmod

**Review by lmoneta — APPROVED** [review]
LGTM!
Thank you Ahmat for this nice improvement!

**Pull request merged**